### DISCUSSION TITLE
Using source generators in future?

### DISCUSSION THREAD

**Discussion by quicoli** [nhibernate/nhibernate-core#3443] [Ideas]
Hi all!

I know... new .NET, a lot of 'noise'... new stuff, new changes... 

After watching David Fowler talk about improvements in .NET, it's noticeable how source generators are highly used now.
Even before that, the [MVVM Community Toolkit](https://learn.microsoft.com/en-us/dotnet/communitytoolkit/mvvm/) 
started using this feature.

I believe that to start the use of Source Generators in NH would be good - performance, project longevity and modernization wise.
I also believe this would be a big effort/code change...

Is there any planning of using Source Generators for future versions?

How could we (I) help with this?

**Comment by fredericDelaporte** [comment]
Hello,

I am not aware of any contributor having plans about source generators.

I can see one probably not so hard win with source generators: proxies generation. Currently proxies are generated at runtime as dynamically generated assemblies. (See the Nhibernate.Proxy namespace, especially the static proxy factory.) Generating them while compiling projects using NHibernate would avoid that runtime cost for generating proxies. (It would not improve the performances of an application once "hot", it would just improve its warm-up.)

The main hurdle for the source generator would be to access and analyze the NHibernate mappings in order to know which proxies have to be generated. There is nothing available currently for doing that in a source generator stage. So, a prerequisite would be to investigate and implement a way for a source generator to access and analyze the mappings, ideally whatever how they are provided (hbm, attributes, by code, convention, ...). Can a source generator run an assembly like the NHibernate one? It basically would have to build the session factory to get all the mapping information.

After that, for what would you attempt to use it? Avoiding reflection in entities hydration? It looks to me that would require some heavy changes in how NHibernate works.

**Reply by quicoli** [reply]
Hi @fredericDelaporte , thank you for this class on how NH works. 

When I first mentioned source generators I was thinking on avoiding reflection and already was wondering this would require some big changes...

In the next weeks I'll look at NH code to understand it better as one thing is to know how to use and other is how it is built :)

Thanks again!

**Comment by hazzik** [comment]
@quicoli there is this project: https://github.com/nhibernate/NHibernate.ProxyGenerators which used to generate proxies for the classes. It uses old code generators, but still works. I think it would be a good candidate to use code generators.

**Comment by fredericDelaporte** [comment]
NHibernate also uses [`AsyncGenerator`](https://github.com/maca88/AsyncGenerator), which currently does not use source generators but adds generated source code files. Evolving this tool into a source generator would remove a lot of generated source files from NHibernate. But that would possibly be a quite heavy source generator slowing down the build a bit too much.

**Comment by danfma** [comment]
The best approach is to use source generators to generate proxy and unsafe accessors for NHibernate, which would greatly raise the performance at runtime.

https://neuecc.medium.com/introducing-the-privateproxy-library-utilizing-net-8-unsafeaccessor-c47246be4b3e
